freebsd/hid.c

```diff
@@ -0,0 +1,1143 @@
+	desc.hgd_report_type = HID_OUTPUT_REPORT;
+
+	res = ioctl(dev->device_handle, HIDRAW_SET_REPORT, &desc);
+	if (res < 0)
+		register_device_error_format(dev, "ioctl (SOUTPUT): %s", strerror(errno));
+
+	return res;
+}
+
+int HID_API_EXPORT HID_API_CALL hid_get_input_report(hid_device *dev, unsigned char *data, size_t length)
+{
+	int res;
+	struct hidraw_gen_descriptor desc;
+
+	if (!data || (length == 0)) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	desc.hgd_maxlen = length;
+	desc.hgd_data = (void *)data;
+	desc.hgd_report_type = HID_OUTPUT_REPORT;
+
+	register_device_error(dev, NULL);
+
+	res = ioctl(dev->device_handle, HIDRAW_GET_REPORT, &desc);
+	if (res < 0)
+		register_device_error_format(dev, "ioctl (GINPUT): %s", strerror(errno));
+
+	return res;
+}
+
+void HID_API_EXPORT hid_close(hid_device *dev)
+{
+	if (!dev)
+		return;
+
+	close(dev->device_handle);
+
+	free((void *)dev->device_path);
+	free(dev->error_str);
+	free(dev->read_error_str);
+
+	hid_free_enumeration(dev->device_info);
+
+	free(dev);
+}
+
+
+int HID_API_EXPORT_CALL hid_get_manufacturer_string(hid_device *dev, wchar_t *string, size_t maxlen)
+{
+	if (!string || !maxlen) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	struct hid_device_info *info = hid_get_device_info(dev);
+	if (!info) {
+		// hid_get_device_info will have set an error already
+		return -1;
+	}
+
+	if (info->manufacturer_string) {
+		wcsncpy(string, info->manufacturer_string, maxlen);
+		string[maxlen - 1] = L'\0';
+	}
+	else {
+		string[0] = L'\0';
+	}
+
+	return 0;
+}
+
+int HID_API_EXPORT_CALL hid_get_product_string(hid_device *dev, wchar_t *string, size_t maxlen)
+{
+	if (!string || !maxlen) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	struct hid_device_info *info = hid_get_device_info(dev);
+	if (!info) {
+		// hid_get_device_info will have set an error already
+		return -1;
+	}
+
+	if (info->product_string) {
+		wcsncpy(string, info->product_string, maxlen);
+		string[maxlen - 1] = L'\0';
+	}
+	else {
+		string[0] = L'\0';
+	}
+
+	return 0;
+}
+
+int HID_API_EXPORT_CALL hid_get_serial_number_string(hid_device *dev, wchar_t *string, size_t maxlen)
+{
+	if (!string || !maxlen) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	struct hid_device_info *info = hid_get_device_info(dev);
+	if (!info) {
+		// hid_get_device_info will have set an error already
+		return -1;
+	}
+
+	if (info->serial_number) {
+		wcsncpy(string, info->serial_number, maxlen);
+		string[maxlen - 1] = L'\0';
+	}
+	else {
+		string[0] = L'\0';
+	}
+
+	return 0;
+}
+
+
+HID_API_EXPORT struct hid_device_info *HID_API_CALL hid_get_device_info(hid_device *dev) {
+	int idx;
+
+	if (dev->device_info) {
+		register_device_error(dev, NULL);
+	}
+	else {
+		// Lazy initialize device_info
+		sscanf(dev->device_path, "/dev/hidraw%d", &idx);
```

Comment:
we set a `device_path` inside of `hid_open_path`, and user may provide a simlink, which may not be in format `/dev/hidrawX`

I'm thinking maybe resolving symlinks would help inside of `hid_open_path`, but I don't know if there could be other special files that are hidraw devices but not located at `/dev/hidrawX` but otherwise a valid hidraw devices...